src/evaluators.js

```diff
@@ -94,29 +91,40 @@ export function createSafeEvaluatorFactory(
       ...mandatoryTransforms
     ];
 
+    // todo (shim limitation): scan endowments, throw error if endowment
+    // overlaps with the const optimization (which would otherwise
+    // incorrectly shadow endowments), or if endowments includes 'eval'. Also
+    // prohibit accessor properties (to be able to consistently explain
+    // things in terms of shimming the global lexical scope).
+    // writeable-vs-nonwritable == let-vs-const, but there's no
+    // global-lexical-scope equivalent of an accessor, outside what we can
+    // explain/spec
+    const scopeHandler = createScopeHandler(
+      unsafeRec,
+      safeGlobal,
+      endowments,
+      sloppyGlobals
+    );
+    const scopeProxy = new Proxy({}, scopeHandler);
+    const scopedEvaluator = apply(scopedEvaluatorFactory, safeGlobal, [
+      scopeProxy
+    ]);
+
     // We use the the concise method syntax to create an eval without a
     // [[Construct]] behavior (such that the invocation "new eval()" throws
     // TypeError: eval is not a constructor"), but which still accepts a
     // 'this' binding.
     const safeEval = {
       eval(src) {
         src = `${src}`;
+
         // Rewrite the source, threading through rewriter state as necessary.
         const rewriterState = allTransforms.reduce(
```

Comment:
I don't get it. Where is `rewriterState.endowments` used? This can't work if it is ignored.

Reply:
Ha... some magic! It's just moved down.

- The `proxyHandler` object holds a closure on the `endowments` object
-`rewriterState.endowments` is also a pointer to the same `endowments` object 
- Therefore, when a rewriter updates an endowment, it's updating `proxyHandler` (othewise tests break).

Bottom-line is that both the proxy and the handler don't have to be regenerated at every evaluation. That's one advantage of not using object inheritance between `endowments` and `safeGlobal` to define the scope resolution. 

Instead, the scope resolution is defined in code, by looking on `endowments` first, then on `safeGlobal` in the proxy traps.